provisioner/template.go

```diff
@@ -582,22 +582,33 @@ func nodeCIDRMaxPods(maskSize int64, extraCapacity int64) (int64, error) {
 	return maxPods, nil
 }
 
-// addressNFromIPv6CIDR takes an IPv6 CIDR and returns the Nth address in the
-// subnet.
-func addressNFromIPv6CIDR(cidr string, n int) (string, error) {
+// nthAddressFromCIDR takes an IPv4 or IPv6 CIDR and returns the Nth address in
+// the subnet.
+func nthAddressFromCIDR(cidr string, n int) (string, error) {
 	_, ipNet, err := net.ParseCIDR(cidr)
 	if err != nil {
 		return "", err
 	}
 
 	ip := ipNet.IP
-	ip = ip.To16()
-	if ip == nil {
-		return "", fmt.Errorf("invalid IP address: %s", ipNet.IP)
+
+	nthByte := 0
+	if ip.To4() != nil {
+		// ipv4 address
+		nthByte = 3
+		ip = ip.To4()
+	} else {
+		// ipv6 address
+		ip = ip.To16()
+		if ip == nil {
+			return "", fmt.Errorf("invalid IP address: %s", ipNet.IP)
+		}
+		nthByte = 15
 	}
+
 	ip = ip.Mask(ipNet.Mask)
 	for i := 0; i < n; i++ {
-		ip[15]++
+		ip[nthByte]++
```

Comment:
@mikkeloscar , looking at this a bit closer I see a few issues. 

For example, if I give the CIDR `192.168.0.0/30`. The current implementation will provide me the `n=10` address as `192.168.0.10` (incrementing the nth byte by `10`) which is incorrect. Since the subnet only really has 4 addresses (192.168.0.0-3). 

I think we need to catch the case where n is potentially > number of host addresses in a subnet so we don't end up in this situation.

(We may never end up in it, but it just came to my mind when looking at the code again..)

Reply:
I think this could be solved like:
```go
package main

import (
	"fmt"
	"log"
	"math/big"
	"net"
)

func main() {
	const (
		cidr = "2a05:d014:9c0:bf05::/64"
		n    = 50
	)

	_, ipNet, err := net.ParseCIDR(cidr)
	if err != nil {
		log.Fatal(err)
	}

	bigN := big.NewInt(int64(n))
	ones, bits := ipNet.Mask.Size()

	if bigN.BitLen() > bits-ones {
		log.Fatalf("n %d requires nBits %d which is greater than 2^%d", n, bigN.BitLen(), bits-ones)
	}

	intIP := new(big.Int)
	intIP.SetBytes(ipNet.IP)
	intIP.Add(intIP, bigN)
	addr := net.IP(intIP.Bytes())

	fmt.Printf("addr: %s\n", addr)
}
```

Reply:
Thanks both!

It definitely makes sense to change this to make it more robust. What I did just resolved what we needed for now, but I'm all for an implementation without edge cases in case we need it for something else in the future.